kvs/src/main.rs

```diff
@@ -120,80 +141,123 @@ where
-                            }
-                        }
-                    }
-                });
+                let mut hf = HydroflowBuilder::default();
+
+                let (z_send, z_recv) = hf.add_channel_input::<_, Option<u64>, VecHandoff<u64>>("ticks");
 
+                // Construct the ticker.
                 let epoch_duration = Duration::from_millis(100);
                 tokio::spawn(async move {
+                    let mut tick = 0;
                     loop {
                         tokio::time::sleep(epoch_duration).await;
-                        epoch += 1;
-                        let my_clock = Single((id, epoch));
-                        // let mut clock = clock.lock().await;
-                        // <ClockRepr as Merge<ClockUpdateRepr>>::merge(&mut clock, Single((id, epoch)));
-                        let mut batches: Vec<<BatchRepr<K, V> as LatticeRepr>::Repr> =
-                            (0..workers).map(|_| Default::default()).collect();
-
-                        for (k, (mut ts, v)) in current_updates.lock().await.drain() {
-                            <ClockRepr as Merge<ClockUpdateRepr>>::merge(&mut ts, my_clock);
-                            // TODO(justin): save a clone here.
-                            for owner in owners(workers, &k) {
-                                <BatchRepr<K, V> as Merge<UpdateRepr<K, V>>>::merge(
-                                    &mut batches[owner],
-                                    Single((k.clone(), (ts.clone(), v.clone()))),
-                                );
-                            }
-                        }
-
-                        // TODO(justin): do this in parallel?
-                        // TODO(justin): reuse the memory by keeping the vecs around?
-                        for (s, batch) in senders.iter().zip(batches.into_iter()) {
-                            s.send(Message::Batch((id, epoch), batch)).await.unwrap();
-                        }
+                        z_send.give(Some(tick));
+                        z_send.flush();
+                        tick += 1;
                     }
                 });
 
-                event_loop.await.unwrap();
+                let (reads_send, reads_recv) =
+                    hf.add_channel_input::<_, Option<K>, VecHandoff<K>>("reads");
+
+                let _ = reads_send;
+
+                let (q_send, q_recv) =
+                    hf.add_channel_input::<_, Option<Message<K, V>>, VecHandoff<Message<K, V>>>("writes");
+                // Feed incoming messages into a Hydroflow channel.
+                // TODO(justin): not sure why the tokio Receiver doesn't support try_iter?
+                tokio::spawn(async move {
+                    // TODO(justin): batch all these per try_recv?
+                    while let Some(v) = receiver.recv().await {
+                        q_send.give(Some(v));
+                        q_send.flush();
+                    }
+                });
```

Comment:
You can use [`ReceiverStream`](https://docs.rs/tokio-stream/0.1.8/tokio_stream/wrappers/struct.ReceiverStream.html) to turn `receiver` into an async stream and use `add_input_from_stream` instead